### PR TITLE
Build offline view mode of iAnime

### PR DESCRIPTION
**Pull request by Shiney-Kelly**
## Work Summary

- Add refresh logic, for first retrieve data from local database only when database is empty or user trigger refresh
- Add get specific anime details from database
- Separate setup for recycler view in fragment
- Reset primary key when db is cleared
- Pass only ID in adapter
- Handle both anime found and not found scenarios in anime detail screen

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Code enhancement (optimize or refactor code)

## How Has This Been Tested?
**Test Screenshots**:
Screenshot 2024-06-02 at 3 26 50 PM

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/src/main/java/com/project/ianime/data/AnimeDao.kt [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    @Query("DELETE FROM ianime_offline_table")
+    suspend fun deleteAll()
+
+    @Query("SELECT COUNT(*) from ianime_offline_table")
+    fun getAnimeCount(): Int
+
+    @Query("SELECT * FROM ianime_offline_table WHERE anime_id = :animeId")
+    fun getAnimeById(animeId: String): Flow<AnimeEntity?>
+
+    @Query("DELETE FROM sqlite_sequence WHERE name= 'ianime_offline_table'")
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/data/AnimeDao.kt]: just small spacing here before `=`: `WHERE name= 'ianime_offline_table'`

File: app/src/main/java/com/project/ianime/screens/gallery/GalleryFragment.kt [hunk 2]
```diff
@@ -78,13 +80,15 @@ class GalleryFragment : BaseFragment() {
         binding.emptyContainer.root.visibility = View.GONE
         loadingBar.visibility = View.GONE
         animeCardList.visibility = View.VISIBLE
+    }
 
+    private fun setUpRecyclerView(){
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/screens/gallery/GalleryFragment.kt]: small spacing before `{`

File: app/src/main/java/com/project/ianime/screens/viewanime/AnimeDetailFragment.kt [hunk 3]
```diff
@@ -78,26 +78,29 @@ class AnimeDetailFragment : BaseFragment() {
...
+                targetAnime.animeImageUrl?.let {
+                    imageUtils.loadImageFromDisk(it, animeProfile)
+                }
+                animeType.text = getString(R.string.anime_type_title, getString(it.type.label))
+                animeCountry.text = getString(R.string.anime_country_title, getString(it.country.label))
+                val releaseYear = it.releaseYear ?: getString(R.string.not_applicable_data)
+                animeReleasedYear.text = getString(R.string.anime_release_year_title, releaseYear)
+                animeStatus.text = getString(R.string.anime_status_title, getString(it.status.label))
+                val animeDescription: String = it.synopsis ?: getString(R.string.empty_description_message)
+                animeIntro.text = animeDescription
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/screens/viewanime/AnimeDetailFragment.kt]: do we need to consider empty value here?

File: app/src/main/java/com/project/ianime/viewmodels/AnimeViewModel.kt [hunk 4]
```diff
@@ -17,35 +23,63 @@ class AnimeViewModel @Inject constructor(private val repository: AnimeDataReposi
...
-     * get entire list of animes with all information
+     * load entire anime list with refresh control
      */
     fun getAnimeList(refresh: Boolean = false) {
-        if (refresh){
-            getAnimeListFromNetwork()
+        viewModelScope.launch {
+            withContext(Dispatchers.IO) {
+                // if refresh or database is empty
+                if (refresh || repository.isDatabaseEmpty()){
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/viewmodels/AnimeViewModel.kt]: `{` spacing

File: app/src/main/java/com/project/ianime/viewmodels/AnimeViewModel.kt [hunk 5]
```diff
@@ -17,35 +23,63 @@ class AnimeViewModel @Inject constructor(private val repository: AnimeDataReposi
...
         repository.getAnimeListFromNetwork()
             .observeOn(AndroidSchedulers.mainThread())
             .subscribe({ animeItems ->
+                Log.d("AnimeViewModel", "Fetched from network: ${animeItems.size} items")
                 if (animeItems.isEmpty()) {
                     animeUiState.postValue(AnimeUiState.Empty)
                 } else {
+                    viewModelScope.launch {
+                        // clear existing data from the database
+                        withContext(Dispatchers.IO){
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/viewmodels/AnimeViewModel.kt]: `{` spacing

File: app/src/main/java/com/project/ianime/viewmodels/AnimeViewModel.kt [hunk 6]
```diff
@@ -17,35 +23,63 @@ class AnimeViewModel @Inject constructor(private val repository: AnimeDataReposi
...
+                            repository.clearOfflineAnimeList()
+                        }
+
+                        // insert new data into the database
+                        withContext(Dispatchers.IO){
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/viewmodels/AnimeViewModel.kt]: `{` spacing

**Review by Tony-Feng — APPROVED** [review]

File: app/src/main/java/com/project/ianime/screens/viewanime/AnimeDetailFragment.kt (hunk 3 above)
Comment (Shiney-Kelly) [reply app/src/main/java/com/project/ianime/screens/viewanime/AnimeDetailFragment.kt]: see [add anime not found scenario](https://github.com/AM-Kitty/iAnime/pull/10/commits/694137e437fade69a0bf36cbb9add7e0fd2a3dd8)

**Review by Tony-Feng — APPROVED** [review]

**Pull request merged**